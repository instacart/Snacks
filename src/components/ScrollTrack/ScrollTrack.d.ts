import * as React from 'react'
import { Interpolation } from '@emotion/core'

interface CallbackProps {
  atStart: boolean
  atEnd: boolean
  slideTo: number
  parentWidth: number
  trackWidth: number
}

export interface ScrollTrackProps {
  /** Prop for passing in custom button content for back button */
  backButtonContent?: React.ReactNode

  /** The elements to scroll */
  children?: React.ReactNode

  className?: string

  /** Manually control left positioning of ScrollTrack */
  leftOverride?: number

  /** Prop for passing in custom button content for next button */
  nextButtonContent?: React.ReactNode

  /**
   * A callback called before sliding to next set.
   * ** Passed function must return a promsie **
   * -- will wait for promise resolution before continuing slide.
   * Use for high levels of control
   */
  onBeforeNext?(callbackProps: CallbackProps): Promise<undefined | { scrollOffset?: number }>

  /**
   * A callback called before sliding to previous set.
   * ** Passed function must return a promsie **
   * -- will wait for promise resolution before continuing slide.
   * Use for high levels of control
   */
  onBeforeBack?(callbackProps: CallbackProps): Promise<undefined | { scrollOffset?: number }>

  /**  function to be called after sliding to next set. */
  onAfterNext?(callbackProps: CallbackProps): void

  /**  function to be called after sliding to previous set. */
  onAfterBack?(callbackProps: CallbackProps): void

  /** number of pixels to offset forward scrolls by */
  scrollOffset?: number

  /** Speed of scrolling animaton in milleseconds - defaults to 150ms */
  scrollSpeed?: number

  /** Transition timing function to use for scrolling animation - defaults to ease-in-out */
  scrollTimingFunction?: string

  /**
   * Optional style overrides merged into emotion css
   *
   * @deprecated
   * This prop exists for backwards compatibility and will be
   * removed in a future version
   */
  style?: Interpolation

  /** Style specifc children elements [LeftArrow, RightArrow, Track] */
  styles?: {
    LeftArrow?: Interpolation
    RightArrow?: Interpolation
    Track?: Interpolation
  }
}

declare const ScrollTrack: React.ComponentClass<ScrollTrackProps>

export default ScrollTrack
