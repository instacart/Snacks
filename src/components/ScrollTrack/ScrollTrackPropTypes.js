import PropTypes from 'prop-types'

const propTypes = {
  trackProps: PropTypes.shape({
    showLeftArrow: PropTypes.bool,
    showRightArrow: PropTypes.bool,
    left: PropTypes.number,
    parentWidth: PropTypes.number,
    trackWidth: PropTypes.number,
    trackBounds: PropTypes.object,
  }),
}

export default propTypes
