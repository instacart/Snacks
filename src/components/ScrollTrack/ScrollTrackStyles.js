import responsive from '../../styles/responsive'
import spacing    from '../../styles/spacing'

export default {
  containerStyles: {
    overflow: 'hidden',
    position: 'relative',
  },
  innerContainerStyles: {
    whiteSpace: 'nowrap',
    position: 'relative',

    [responsive.xs]: {
      left: 'inherit',
      overflowX: 'scroll',
      display: 'block',
      msOverflowStyle: 'none',
      overflow: '-moz-scrollbars-none',
      webkitOverflowScrolling: 'touch'
    },
  },
  slideButtonStyles: {
    default: {
      position: 'absolute',
      top: '5px',
    },
    left: spacing.LEFT_XS,
    right: spacing.RIGHT_XS
  }
}
