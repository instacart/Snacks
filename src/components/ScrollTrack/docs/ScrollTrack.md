### ScollTrack example:

    internalTrackStyles = {
      RightArrow: {
        backgroundColor: 'blue'
      }
    }

    styles = { padding: '8px 20px', marginRight: '5px', fontSize: '24px', borderRadius: '4px', backgroundColor: '#eee', color: '#43B02A', top: '8px' };
    <div style={{ height: '56px' }}>
      <ScrollTrack styles={internalTrackStyles}>
        <div style={{ padding: '8px 0' }}>
          <Icon name="cart" style={styles} />
          <Icon name="cart" style={styles} />
          <NavigationPill
            isActive={true}
            onClick={e => {
              e.preventDefault()
              console.log('NavigationPill click!', "bananas")
            }}
            text={"bananas"}
          />
          <Icon name="cart" style={styles} />
          <Icon name="cart" style={styles} />
          <NavigationPill
            isActive={true}
            onClick={e => {
              e.preventDefault()
              console.log('NavigationPill click!', "carrots")
            }}
            text={"carrots"}
          />
          <Icon name="cart" style={styles} />
          <Icon name="cart" style={styles} />
          <NavigationPill
            isActive={true}
            onClick={e => {
              e.preventDefault()
              console.log('NavigationPill click!', "watermelon")
            }}
            text={"watermelon"}
          />
          <Icon name="cart" style={styles} />
          <Icon name="cart" style={styles} />
          <NavigationPill
            isActive={true}
            onClick={e => {
              e.preventDefault()
              console.log('NavigationPill click!', "kale")
            }}
            text={"kale"}
          />
          <Icon name="cart" style={styles} />
          <Icon name="cart" style={styles} />
          <NavigationPill
            isActive={true}
            onClick={e => {
              e.preventDefault()
              console.log('NavigationPill click!', "arugula")
            }}
            text={"arugula"}
          />
          <Icon name="cart" style={styles} />
          <Icon name="cart" style={styles} />
          <NavigationPill
            isActive={true}
            onClick={e => {
              e.preventDefault()
              console.log('NavigationPill click!', "spinach")
            }}
            text={"spinach"}
          />
        </div>
      </ScrollTrack>
    </div>

### ScrollTrack with equalWidthTrack wrapper
**NOTE: This HOC assumes the wrapped component is a container of equally and static sized elements**

`ScrollTrack.equalWidthTrack` is a higher order component that will supply the wrapped
component with two additional props: `startIndex` and `lastIndex`. This will enable the
child to determine how it wants to treat elements that are shown within the track
or hidden off the overflow.

Example use case: *A container of item cards that all have a width of 90px and margins of 5px on both sides*

```js static
  const childWidth = 100;

  @ScrollTrack.equalWidthTrack(childWidth)
  class ItemCardsContainer extends Component {
    static propTypes = {
      trackProps: ScrollTrack.ScrollTrackPropTypes.trackProps,
      startIndex: PropTypes.number,
      lastIndex: PropTypes.number
    };

    render () {
      ...
    }
  }
```

    const CustomComponent = ScrollTrack.equalWidthTrack(186)(({ startIndex, lastIndex, trackProps }) => {
      const { showRightArrow, showLeftArrow } = trackProps;
      const children = new Array(20).fill(1).map((_, index) => {
        const style = {
          backgroundColor: 'red',
          minWidth: '176px',
          minHeight: '176px',
          margin: '0 5px',
          opacity: (showLeftArrow && index === startIndex) || (showRightArrow && index === lastIndex) ? 0.6 : 1
        };

        return <div style={style}>Item</div>;
      });

      return (
        <div style={{ display: 'flex' }}>
          {children}
        </div>
      );
    });

    CustomComponent.propTypes = {
      trackProps: ScrollTrack.ScrollTrackPropTypes.trackProps,
    };

    <ScrollTrack>
      <CustomComponent />
    </ScrollTrack>


Using callbacks
```js static
  <ScrollTrack
    onBeforeNext={(props) => {
      return new Promise((resolve, reject) => {
          setTimeout(() => {
            console.log('before next!', props)
            resolve()
          }, 1000)
      })
    }}
    onAfterNext={(props) => { console.log('after next!', props) }}
    onBeforeBack={(props) => { console.log('before back!', props) }}
    onAfterBack={(props) => { console.log('after back!', props) }}
  >
    <div>
      <Icon name="cart" style={styles} />
      <Icon name="cart" style={styles} />
      <Icon name="cart" style={styles} />
      <Icon name="cart" style={styles} />
      <Icon name="cart" style={styles} />
      <Icon name="cart" style={styles} />
    </div>
  </ScrollTrack>
```

Using custom next and back buttons

    <ScrollTrack
      backButtonElement={(
        <button>
          <Icon name="arrowLeft" />
        </button>
      )}
      nextButtonElement={(
        <button>
          <Icon name="arrowRight" />
        </button>
      )}
      styles={{
          Track: { height: '56px' },
          LeftArrow: { backgroundColor: '#7FE364', color: '#fff', border: 'none', borderRadius: '50%', height: '30px', width: '30px', lineHeight: '30px', fontWeight: '600' },
          RightArrow: { backgroundColor: '#7FE364', color: '#fff', border: 'none', borderRadius: '50%', height: '30px', width: '30px', lineHeight: '30px', fontWeight: '600' }
      }}
    >
      <div>
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
        <Icon name="cart" style={styles} />
      </div>
    </ScrollTrack>
