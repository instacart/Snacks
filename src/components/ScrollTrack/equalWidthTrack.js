/* eslint-disable max-classes-per-file */
import React, { PureComponent } from 'react'
import ScrollTrackPropTypes from './ScrollTrackPropTypes'

class EqualWidthTrackError extends TypeError {
  name = 'EqualWidthTrackError'
}

/**
 * A higher order component that will supply the wrapped
 * component with two additional props: `startIndex` and `lastIndex`. This will enable the
 * child to determine how it wants to treat elements that are shown within the track
 * or hidden off the overflow.
 */
const equalWidthTrack = childWidth => {
  if (!['number', 'function'].includes(typeof childWidth)) {
    throw new EqualWidthTrackError('childWidth must be a number or function bassed on props')
  }

  return WrappedComponent => {
    class EqualWidthTrack extends PureComponent {
      static propTypes = {
        trackProps: ScrollTrackPropTypes.trackProps,
      }

      render() {
        if (!this.props.trackProps) {
          return null
        }
        const childWidthNumber =
          typeof childWidth === 'function' ? childWidth(this.props) : childWidth

        const { left, parentWidth } = this.props.trackProps

        const startIndex = Math.floor(Math.abs(left) / childWidthNumber)
        const lastIndex = Math.floor((Math.abs(left) + parentWidth) / childWidthNumber)

        return <WrappedComponent {...this.props} startIndex={startIndex} lastIndex={lastIndex} />
      }
    }

    // Access component that is being wrapped
    EqualWidthTrack.WrappedComponent = WrappedComponent.WrappedComponent
      ? WrappedComponent.WrappedComponent
      : WrappedComponent

    return EqualWidthTrack
  }
}

export default equalWidthTrack
