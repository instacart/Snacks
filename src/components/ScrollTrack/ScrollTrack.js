import componentStyles from './ScrollTrackStyles'
import equalWidthTrack from './equalWidthTrack'
import ScrollTrackPropTypes from './ScrollTrackPropTypes'

import React, { Component } from 'react'
import CircleButton  from '../Buttons/CircleButton'
import Icon          from '../Icon/Icon'
import Radium        from 'radium'
import PropTypes     from 'prop-types'
import _             from 'underscore'

const noOp = () => {} // eslint-disable-line no-empty-function

@Radium
class ScrollTrack extends Component {
  static equalWidthTrack = equalWidthTrack
  static ScrollTrackPropTypes = ScrollTrackPropTypes

  static propTypes = {
    /** Prop for passing in custom button content for back button */
    backButtonContent: React.PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.Symbol,
    ]),

    /** Manually control left positioning of ScrollTrack */
    leftOverride: PropTypes.number,

    /** Prop for passing in custom button content for next button */
    nextButtonContent: React.PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.Symbol,
    ]),

    /**
    * A callback called before sliding to next set.
    * ** Passed function must return a promsie **
    * -- will wait for promise resolution before continuing slide.
    * Use for high levels of control
    */
    onBeforeNext: PropTypes.func,

    /**  function to be called before sliding to previous set. */
    onBeforeBack: PropTypes.func,

    /**  function to be called after sliding to next set. */
    onAfterNext: PropTypes.func,

    /**  function to be called after sliding to previous set. */
    onAfterBack: PropTypes.func,

    /** Transition timing function to use for scrolling animation - defaults to ease-in-out */
    scrollTimingFunction: PropTypes.string,

    /** Speed of scrolling animaton in milleseconds - defaults to 150ms */
    scrollSpeed: PropTypes.number,

    /** Style top level element */
    style: PropTypes.object,

    /** Style specifc children elements [LeftArrow, RightArrow, Track] */
    styles: PropTypes.shape({
      LeftArrow: PropTypes.object,
      RightArrow: PropTypes.object,
      Track: PropTypes.object,
    }),
  }

  static defaultProps = {
    leftOverride: 0,
    scrollSpeed: 150,
    scrollTimingFunction: 'ease-in-out',
    styles: {
      LeftArrow: {},
      RightArrow: {},
      Track: {}
    },
    style: {},
    onBeforeBack: noOp,
    onAfterNext: noOp,
    onAfterBack: noOp,
    onBeforeNext: () => new Promise(resolve => resolve())
  }

  constructor(props) {
    super(props)

    this.state = {
      showLeftArrow: false,
      showRightArrow: false,
      left: props.leftOverride
    }
  }

  componentDidMount() {
    this.debouncdComputeSlideAttributes = _.debounce(this.computeSlideAttributes, 200)
    this.computeSlideAttributes()

    window.addEventListener('resize', this.debouncdComputeSlideAttributes)
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.leftOverride !== this.props.leftOverride) {
      // this allows for control of the scrolltrack by parent components
      this.setState({ left: nextProps.leftOverride })
      this.computeSlideAttributes()
    }
  }

  componentDidUpdate(prevProps) {
    const prevChildren = prevProps.children || []
    const newChildren = this.props.children || []

    if (!_.isEqual(prevChildren, newChildren)) {
      this.computeSlideAttributes()
    }
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.debouncdComputeSlideAttributes)
  }

  get childrenWithTrackProps() {
    const nodeWidths = this.getNodeWidths()
    const trackProps = {
      ...this.state,
      ...nodeWidths
    }

    return React.Children.map(this.props.children, child => {
      const isHtmlTag = typeof child.type === 'string' && child.type[0] === child.type[0].toLowerCase()
      const childProps = isHtmlTag ? {} : { trackProps }

      return React.cloneElement(child, childProps)
    })
  }

  getNodeWidths = () => {
    const parentNode = this.refs.container
    const parentNodeBounds = parentNode && parentNode.getBoundingClientRect()
    const trackNode = this.refs.track
    const trackNodeBounds = trackNode && trackNode.getBoundingClientRect()
    const parentWidth = parentNodeBounds && parentNodeBounds.width
    const trackWidth = trackNode && (trackNode.offsetLeft + trackNode.scrollWidth)
    const trackBounds = trackNodeBounds

    return { parentWidth, trackWidth, trackBounds }
  }

  computeSlideAttributes = () => {
    const { parentWidth, trackWidth } = this.getNodeWidths()
    const trackAtEnd = parentWidth < trackWidth && this.state.left <= (parentWidth - trackWidth)
    const trackAtBeginning = this.state.left >= 0

    if (!parentWidth || !trackWidth) { return }
    if (Math.ceil(parentWidth) >= trackWidth) { return this.hideArrows() }
    if (!trackAtEnd) { this.showRightArrow() } else { this.hideRightArrow() }
    if (!trackAtBeginning) { this.showLeftArrow() } else { this.hideLeftArrow() }
  }

  hideArrows = () => {
    this.setState({
      showLeftArrow: false,
      showRightArrow: false
    })
  }

  hideRightArrow = () => {
    this.setState({ showRightArrow: false })
  }

  hideLeftArrow = () => {
    this.setState({ showLeftArrow: false })
  }

  showRightArrow = () => {
    this.setState({ showRightArrow: true })
  }

  showLeftArrow = () => {
    this.setState({ showLeftArrow: true })
  }

  slideForward = () => {
    const { parentWidth, trackWidth } = this.getNodeWidths()
    let nextForward = this.state.left - parentWidth
    const fullForward = parentWidth - trackWidth
    const { onBeforeNext, onAfterNext } = this.props

    // already is, or is going to be, full forward
    if (nextForward <= fullForward) { nextForward = fullForward }

    const callbackProps = {
      atStart: trackWidth <= parentWidth,
      atEnd: fullForward === nextForward,
      slideTo: nextForward,
      parentWidth,
      trackWidth
    }

    onBeforeNext(callbackProps).then(() => {
      this.updateLeftValue({
        left: nextForward,
        callback: onAfterNext,
        callbackProps
      })
    })
  }

  slideBack = () => {
    const { parentWidth, trackWidth } = this.getNodeWidths()
    let nextBack = this.state.left + parentWidth
    const { onBeforeBack, onAfterBack } = this.props

    // already is, or is going to be, full back
    if (this.state.left >= 0 || nextBack >= 0) { nextBack = 0 }

    const callbackProps = {
      atStart: nextBack === 0,
      atEnd: false,
      slideTo: nextBack,
      parentWidth,
      trackWidth
    }

    onBeforeBack(callbackProps)

    this.updateLeftValue({
      left: nextBack,
      callback: onAfterBack,
      callbackProps
    })
  }

  updateLeftValue({left, callback, callbackProps}) {
    this.setState({ left }, () => {
      this.computeSlideAttributes()
      callback(callbackProps)
    })
  }

  renderRightArrow = () => {
    const { slideButtonStyles } = componentStyles
    const { showRightArrow } = this.state
    const { styles: { RightArrow = {} }, nextButtonContent } = this.props

    return (
      <CircleButton
        onClick={this.slideForward}
        ariaLabel='next'
        styles={[
          slideButtonStyles.default,
          slideButtonStyles.right,
          showRightArrow && { display: 'block' },
          RightArrow
        ]}
      >
        { nextButtonContent ||
          <Icon
            name='arrowRightSmallBold'
            style={{ fontSize: '20px' }}
          />
        }
      </CircleButton>
    )
  }

  renderLeftArrow = () => {
    const { slideButtonStyles } = componentStyles
    const { showLeftArrow } = this.state
    const { styles: { LeftArrow = {} }, backButtonContent } = this.props

    return (
      <CircleButton
        onClick={this.slideBack}
        ariaLabel='back'
        styles={[
          slideButtonStyles.default,
          slideButtonStyles.left,
          showLeftArrow && { display: 'block' },
          LeftArrow
        ]}
      >
        { backButtonContent ||
          <Icon
            name='arrowLeftSmallBold'
            style={{ fontSize: '20px' }}
          />
        }
      </CircleButton>
    )
  }

  render() {
    const { containerStyles, innerContainerStyles } = componentStyles
    const { children, scrollSpeed, scrollTimingFunction, style, styles: { Track = {} } } = this.props

    if (!children) { return null }

    return (
      <div
        ref='container'
        style={{ ...containerStyles, ...style }}
      >
        {this.renderLeftArrow()}
        <div
          style={[
            {
              transition: `transform ${scrollSpeed}ms ${scrollTimingFunction}`,
              transform: `translate3d(${this.state.left}px, 0, 0)`
            },
            innerContainerStyles
          ]}
        >
          <div ref='track' style={Track}>
            {this.childrenWithTrackProps}
          </div>
        </div>
        {this.renderRightArrow()}
      </div>
    )
  }
}


export default ScrollTrack
