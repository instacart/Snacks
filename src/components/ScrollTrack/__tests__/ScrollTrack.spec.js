import React from 'react'
import renderer from 'react-test-renderer'
import ScrollTrack from '../ScrollTrack'
import { StyleRoot } from 'radium'
import { ReactWrapper, mount } from 'enzyme'
import { spy } from 'sinon'

it('renders ScrollTrack correctly', () => {
  const tree = renderer.create(
    <StyleRoot>
      <div>
        <ScrollTrack leftOverride={0} >
          <p>one</p>
          <p>two</p>
          <p>three</p>
          <p>four</p>
          <p>five</p>
          <p>six</p>
          <p>seven</p>
        </ScrollTrack>
      </div>
    </StyleRoot>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders ScrollTrack without children correctly', () => {
  const tree = renderer.create(
    <StyleRoot>
      <div>
        <ScrollTrack leftOverride={0} >
        </ScrollTrack>
      </div>
    </StyleRoot>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders ScrollTrack buttons correctly', () => {
  const styles = {
    RightArrow: {
      backgroundColor: 'blue'
    },
    LeftArrow: {
      backgroundColor: 'red'
    },
  }

  const track = mount(
    <StyleRoot>
      <div>
        <ScrollTrack leftOverride={0} styles={styles}>
          <p>one</p>
          <p>two</p>
          <p>three</p>
          <p>four</p>
          <p>five</p>
          <p>six</p>
          <p>seven</p>
        </ScrollTrack>
      </div>
    </StyleRoot>
  )

  // should have 2 buttons
  let buttons = track.find('button')
  let leftButton = buttons.first()
  let rightButton = buttons.last()
  expect(buttons).toHaveLength(2)

  // check that neither are showing
  let leftButtonStyle = leftButton.props().style
  let rightButtonStyle = rightButton.props().style
  expect(leftButtonStyle.display).toEqual('none')
  expect(rightButtonStyle.display).toEqual('none')

  // show right arrow
  track.find(ScrollTrack).node.showRightArrow()

  // check arrow is showing and has correct styles
  buttons = track.find('button')
  leftButton = buttons.first()
  rightButton = buttons.last()
  leftButtonStyle = leftButton.props().style
  rightButtonStyle = rightButton.props().style

  // make sure styles are correct
  expect(buttons).toHaveLength(2)
  expect(rightButtonStyle.backgroundColor).toEqual(styles.RightArrow.backgroundColor)
  expect(rightButtonStyle.display).toEqual('block')
  expect(leftButtonStyle.display).toEqual('none')

  // show left arrow
  track.find(ScrollTrack).node.showLeftArrow()
  buttons = track.find('button')
  leftButton = buttons.first()
  rightButton = buttons.last()
  leftButtonStyle = leftButton.props().style
  rightButtonStyle = rightButton.props().style

  // make sure styles are correct
  expect(rightButtonStyle.display).toEqual('block')
  expect(leftButtonStyle.display).toEqual('block')
  expect(leftButtonStyle.backgroundColor).toEqual(styles.LeftArrow.backgroundColor)

  // hide both arrows
  track.find(ScrollTrack).node.hideArrows()
  buttons = track.find('button')
  leftButton = buttons.first()
  rightButton = buttons.last()
  leftButtonStyle = leftButton.props().style
  rightButtonStyle = rightButton.props().style

  // ensure both are hidden
  expect(leftButtonStyle.display).toEqual('none')
  expect(rightButtonStyle.display).toEqual('none')
})

it('renders custom ScrollTrack buttons correctly', () => {
  const styles = {
    RightArrow: {
      backgroundColor: 'blue'
    },
    LeftArrow: {
      backgroundColor: 'red'
    },
  }

  const track = mount(
    <StyleRoot>
      <div>
        <ScrollTrack
          styles={styles}
          backButtonElement={(
            <button>
              <span>back</span>
            </button>
          )}
          nextButtonElement={(
            <button>
              <span>next</span>
            </button>
          )}
        >
          <p>one</p>
          <p>two</p>
          <p>three</p>
          <p>four</p>
          <p>five</p>
          <p>six</p>
          <p>seven</p>
        </ScrollTrack>
      </div>
    </StyleRoot>
  )

  // should have 2 buttons
  let buttons = track.find('button')
  let leftButton = buttons.first()
  let rightButton = buttons.last()
  expect(buttons).toHaveLength(2)

  // check that neither are showing
  let leftButtonStyle = leftButton.props().style
  let rightButtonStyle = rightButton.props().style
  expect(leftButtonStyle.display).toEqual('none')
  expect(rightButtonStyle.display).toEqual('none')

  // show right arrow
  track.find(ScrollTrack).node.showRightArrow()

  // check arrow is showing and has correct styles
  buttons = track.find('button')
  leftButton = buttons.first()
  rightButton = buttons.last()
  leftButtonStyle = leftButton.props().style
  rightButtonStyle = rightButton.props().style

  // make sure styles are correct
  expect(buttons).toHaveLength(2)
  expect(rightButtonStyle.backgroundColor).toEqual(styles.RightArrow.backgroundColor)
  expect(rightButtonStyle.display).toEqual('block')
  expect(leftButtonStyle.display).toEqual('none')

  // show left arrow
  track.find(ScrollTrack).node.showLeftArrow()
  buttons = track.find('button')
  leftButton = buttons.first()
  rightButton = buttons.last()
  leftButtonStyle = leftButton.props().style
  rightButtonStyle = rightButton.props().style

  // make sure styles are correct
  expect(rightButtonStyle.display).toEqual('block')
  expect(leftButtonStyle.display).toEqual('block')
  expect(leftButtonStyle.backgroundColor).toEqual(styles.LeftArrow.backgroundColor)

  // hide both arrows
  track.find(ScrollTrack).node.hideArrows()
  buttons = track.find('button')
  leftButton = buttons.first()
  rightButton = buttons.last()
  leftButtonStyle = leftButton.props().style
  rightButtonStyle = rightButton.props().style

  // ensure both are hidden
  expect(leftButtonStyle.display).toEqual('none')
  expect(rightButtonStyle.display).toEqual('none')
})

it('renders ScrollTrack with animation props correctly', () => {
  const tree = renderer.create(
    <StyleRoot>
      <div>
        <ScrollTrack
          scrollTimingFunction={'ease-in'}
          scrollSpeed={1000}
        >
          <p>one</p>
          <p>two</p>
          <p>three</p>
          <p>four</p>
          <p>five</p>
          <p>six</p>
          <p>seven</p>
        </ScrollTrack>
      </div>
    </StyleRoot>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

it('onBefore promises & callbacks called correctly', async () => {
  const onBeforeNext = spy()
  const onBeforeBack = spy()
  const track = mount(
    <StyleRoot>
      <div>
        <ScrollTrack
          onBeforeNext={() => {
            return new Promise((resolve, reject) => {
                setTimeout(() => {
                  onBeforeNext()
                  resolve()
                }, 1)
            })
          }}
          onBeforeBack={() => { onBeforeBack() }}
        >
          <p>one</p>
          <p>two</p>
          <p>three</p>
          <p>four</p>
          <p>five</p>
          <p>six</p>
          <p>seven</p>
        </ScrollTrack>
      </div>
    </StyleRoot>
  )

  // show and click next
  track.find(ScrollTrack).node.showRightArrow()
  track.find('button').last().simulate('click')
  await expect(onBeforeNext.calledOnce).resolves.toBeTruthy
  await expect(onBeforeNext.calledWith({})).resolves.toBeTruthy

  // show and click back
  track.find(ScrollTrack).node.hideRightArrow()
  track.find(ScrollTrack).node.showLeftArrow()
  track.find('button').first().simulate('click')
  await expect(onBeforeBack.calledOnce).resolves.toBeTruthy
  await expect(onBeforeBack.calledWith({})).resolves.toBeTruthy
})

it('onAfter promises & callbacks called correctly', async () => {
  const onBeforeNext = spy()
  const onBeforeBack = spy()
  const onAfterNext = spy()
  const onAfterBack = spy()
  const track = mount(
    <StyleRoot>
      <div>
        <ScrollTrack
          onBeforeBack={onBeforeBack}
          onAfterBack={onAfterBack}
          onAfterNext={onAfterNext}
          onBeforeNext={() => {
            return new Promise((resolve, reject) => {
                setTimeout(() => {
                  onBeforeNext()
                  resolve()
                }, 1)
            })
          }}
        >
          <p>one</p>
          <p>two</p>
          <p>three</p>
          <p>four</p>
          <p>five</p>
          <p>six</p>
          <p>seven</p>
        </ScrollTrack>
      </div>
    </StyleRoot>
  )

  // show and click next
  track.find(ScrollTrack).node.showRightArrow()
  track.find('button').last().simulate('click')
  expect(onAfterNext.calledOnce).toBeTruthy
  expect(onAfterNext.calledWith({})).toBeTruthy
  await expect(onAfterBack.calledAfter(onBeforeNext)).resolves.toBeTruthy

  // show and click back
  track.find(ScrollTrack).node.hideRightArrow()
  track.find(ScrollTrack).node.showLeftArrow()
  track.find('button').first().simulate('click')
  expect(onAfterBack.calledOnce).toBeTruthy
  expect(onAfterBack.calledWith({})).toBeTruthy
  expect(onAfterBack.calledAfter(onBeforeBack)).toBeTruthy
})

it('works without any callbacks passed in', async () => {
  const track = mount(
    <StyleRoot>
      <div>
        <ScrollTrack>
          <p>one</p>
          <p>two</p>
          <p>three</p>
          <p>four</p>
          <p>five</p>
          <p>six</p>
          <p>seven</p>
        </ScrollTrack>
      </div>
    </StyleRoot>
  )

  // show and click next
  track.find(ScrollTrack).node.showRightArrow()
  track.find('button').last().simulate('click')

  // show and click back
  track.find(ScrollTrack).node.hideRightArrow()
  track.find(ScrollTrack).node.showLeftArrow()
  track.find('button').first().simulate('click')
})
