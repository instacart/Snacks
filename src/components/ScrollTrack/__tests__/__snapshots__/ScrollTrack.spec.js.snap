// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ScrollTrack correctly 1`] = `
<div
  data-radium={true}
>
  <div>
    <div
      data-radium={true}
      style={
        Object {
          "overflow": "hidden",
          "position": "relative",
        }
      }
    >
      <div
        className="rmq-16b35912"
        data-radium={true}
        style={
          Object {
            "left": 0,
            "position": "relative",
            "transition": "left 150ms ease-in-out",
            "whiteSpace": "nowrap",
          }
        }
      >
        <div
          data-radium={true}
          style={Object {}}
        >
          <p>
            one
          </p>
          <p>
            two
          </p>
          <p>
            three
          </p>
          <p>
            four
          </p>
          <p>
            five
          </p>
          <p>
            six
          </p>
          <p>
            seven
          </p>
        </div>
      </div>
    </div>
  </div>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "@media (max-width: 767px){ .rmq-16b35912{left: inherit !important;
      overflow-x: scroll !important;
      display: block !important;
      -ms-overflow-style: none !important;
      overflow: -moz-scrollbars-none !important;
      webkit-overflow-scrolling: touch !important;}}",
      }
    }
  />
</div>
`;

exports[`renders ScrollTrack without children correctly 1`] = `
<div
  data-radium={true}
>
  <div />
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
  />
</div>
`;
